### PR TITLE
fix(challenges): harden EKS node autoscaling against silent failures

### PR DESCRIPTION
**Pull request by RishabhJain2018**
## Context

Follow-up to #5159. That PR made cross-account autoscaling *work*; this one makes it hard to break again.

While debugging a production incident where challenge 2319's nodegroup sat at zero nodes with 19 pods `Pending` for 19 days, the cross-account IAM fix in #5159 turned out to be one of several independent gaps. Each one on its own is enough to strand a challenge silently.

## What was wrong

**The cross-account role was never provisioned.** Challenges with `use_host_credentials` keep their EKS cluster in the host's own AWS account, which the autoscale Lambda's execution role cannot reach. #5159 taught the Lambda to assume a role there, but nothing in the codebase ever *creates* that role — it had to be made by hand. Every future host-credentials challenge would have repeated the same incident.

**Failures were invisible.** The handler caught every error and returned a `5xx` body. An async invocation records that as a **success**: no retry, no DLQ, no `Errors` metric, nothing to alarm on. This is precisely why the breakage sat unnoticed for weeks.

**Nothing reconciled.** Autoscaling was purely event-driven and fire-and-forget. One dropped invoke left the nodegroup pinned at zero with no path to recovery. `scripts/monitoring/auto_scale_eks_nodes.py` isn't a safety net — it needs a hand-maintained `~/prod_eks_auth_tokens.json` allowlist and uses global `EKS_AWS_*` credentials, so cross-account challenges were never covered.

**`maxSize` escaped the challenge's cap.** It was `max(target, pending_submissions)`, so 50 pending submissions on a `max_worker_instance=2` challenge set `maxSize=50` — in the host's own account, on the host's bill.

**Nodegroup selection was non-deterministic.** The Lambda took `list_nodegroups()[0]`.

**Misconfiguration was silent.** `use_host_credentials` without `aws_account_id` fell through to the Lambda's own account and produced a confusing `AccessDenied` against a cluster living elsewhere.

**Deployment was manual.** The function's code, env vars, DLQ, and schedule lived only in the AWS console. That's how `EKS_AUTOSCALE_DEFAULT_HOST_ACCOUNT_ID` came to be set to a variable the code never read.

## Changes

| Area | Change |
|---|---|
| `aws_utils.py` | `setup_eks_autoscale_cross_account_role()` provisions the role during cluster setup with the host IAM client already in hand. Idempotent — refreshes the trust policy when the role exists. Best-effort; never aborts cluster setup. |
| `auto_scale_eks_nodes_lambda.py` | Raises `AutoscaleError` on unrecoverable failures so async retry + DLQ + `Errors` actually fire. Adds `{"sweep": true}` reconciliation mode. Clamps `maxSize` to `scale_up_cap`. Prefers the recorded nodegroup name. Warns on missing `aws_account_id`. |
| `views.py` / `urls.py` | `autoscale_meta` returns `nodegroup_name`; new `autoscale_eligible_challenges/` endpoint backs the sweep. |
| `models.py` + migration `0128` | `ChallengeEvaluationCluster.nodegroup_name`, recorded at nodegroup creation. |
| `settings/common.py` | `EKS_AUTOSCALE_LAMBDA_ROLE_ARN`, role/policy names, policy document. |
| `deploy_auto_scale_eks_nodes_lambda.sh` | Single source of truth for code, env, DLQ, and the EventBridge sweep schedule. |
| `scaling-guidelines.md` | Documents both modes, the cross-account setup, and its two prerequisites. |

Behaviour for same-account challenges is unchanged.

## Testing

- **19/19** Lambda unit tests pass locally, including new coverage for the `maxSize` clamp, raise-don't-return, sweep success/partial-failure, recorded nodegroup name, and the missing-`aws_account_id` warning.
- 5 new tests for cross-account role provisioning; 5 for the sweep endpoint and `nodegroup_name` in `autoscale_meta`.
- `black`/`isort` clean; `flake8` diffed against `master` — **zero** new violations.
- Django-side tests were **not** run locally (Django 2.2 needs Python ≤3.9 plus Postgres) and are left to CI. They were reviewed statically: migration dependency `0127`, field parity with the model, no URL shadowing, and the `challengeevaluationcluster` reverse accessor all verified.

## Deploying

`EKS_AUTOSCALE_LAMBDA_ROLE_ARN` must be set on the Django server, otherwise role provisioning is skipped with a warning and new host-credentials challenges still need the role created by hand. Existing clusters aren't retrofitted — their roles were already created manually during the incident.

```bash
ENVIRONMENT=production EVALAI_API_SERVER=https://eval.ai LAMBDA_AUTH_TOKEN=xxx DLQ_ARN=... \
  ./scripts/lambda/deploy_auto_scale_eks_nodes_lambda.sh
```

Worth adding a CloudWatch alarm on the function's `Errors` metric — with this change it finally reports something.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added an autoscale-eligible challenges endpoint for the autoscaling Lambda.
  * Added scheduled “sweep” reconciliation support for EKS node autoscaling.
  * Extended autoscaling metadata to include the evaluation cluster’s recorded node group name.
  * Enabled cross-account autoscaling for host-credential challenges with configurable role/policy provisioning.
* **Bug Fixes**
  * Improved failure handling so unrecoverable autoscaling issues trigger retry/DLQ behavior.
  * Made node group name persistence and cross-account role setup best-effort (non-blocking).
* **Documentation**
  * Updated EKS node autoscaling scaling guidelines and runbook.
* **Tests**
  * Added/updated unit tests for endpoint behavior, sweep/error control flow, and role setup edge cases.

### CODE REVIEW

File: apps/challenges/aws_utils.py [hunk 1]
```diff
@@ -2383,6 +2394,98 @@ def create_eks_nodegroup(challenge, cluster_name):
...
+        return None
+
+    logger.info(
+        "Cross-account autoscale role ready for challenge %s: %s",
+        challenge_obj.pk,
+        role_arn,
+    )
+    return role_arn
+
+
```
[the remark this answers is not in the document]
Comment (RishabhJain2018) [reply apps/challenges/aws_utils.py:2548]: Confirmed and fixed in 17c80e87. Verified `WaiterError.__mro__` is `WaiterError → BotoCoreError → Exception` — not a `ClientError`, so it did escape.

Fixed in two places rather than one:
- Added `except BotoCoreError` to the function itself, which covers `WaiterError` plus the rest of the botocore family (`NoCredentialsError`, `EndpointConnectionError`). The docstring makes the best-effort promise, so the guarantee belongs there.
- Also wrapped the call site as you suggested, so an unanticipated exception type still cannot reach the `ChallengeEvaluationCluster` save.

Added `test_returns_none_when_role_waiter_times_out` and `test_setup_eks_cluster_survives_autoscale_role_failure` — the latter asserts the serializer save and subnet creation still happen when role provisioning blows up.

Note the ast-grep `use jsonify instead of json.dumps` hints on this hunk are false positives: `jsonify` is a Flask helper, and `json.dumps` is the correct way to serialise an IAM policy document for boto3.

File: scripts/lambda/auto_scale_eks_nodes_lambda.py [hunk 2]
```diff
@@ -277,3 +323,84 @@ def handler(event, context):
...
+            )
+        )
+
+    return {
+        "statusCode": 200,
+        "body": json.dumps(
+            {"swept": len(succeeded), "challenge_pks": succeeded}
+        ),
+    }
+
```
[the remark this answers is not in the document]
Comment (RishabhJain2018) [reply scripts/lambda/auto_scale_eks_nodes_lambda.py:390]: Correct, fixed in 17c80e87. `pending_submissions = int(...)` and `scale_up_cap = int(...)` sit outside any try block, so malformed API data raises a bare `ValueError` that would skip every remaining challenge.

Took the broad `except Exception` you suggested, with a comment explaining why it is deliberate here — per-item isolation is the entire purpose of a reconciliation sweep. Added `test_sweep_isolates_non_autoscale_errors` covering the `ValueError` case specifically.

File: scripts/lambda/deploy_auto_scale_eks_nodes_lambda.sh [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    --description "Reconciles EvalAI EKS nodegroup scaling for all eligible challenges" >/dev/null
+
+# Idempotent: the statement may already exist from a previous deploy.
+aws_cli lambda add-permission \
+    --function-name "${FUNCTION_NAME}" \
+    --statement-id "${RULE_NAME}-invoke" \
+    --action "lambda:InvokeFunction" \
+    --principal events.amazonaws.com \
+    --source-arn "$(aws_cli events describe-rule --name "${RULE_NAME}" \
+        --query "Arn" --output text)" >/dev/null 2>&1 || true
```
[the remark this answers is not in the document]
Comment (RishabhJain2018) [reply scripts/lambda/deploy_auto_scale_eks_nodes_lambda.sh]: Correct, fixed in 17c80e87. Fair catch given this PR argues for making failures visible.

Used your approach, hoisting the `describe-rule` call out of the command substitution first so a failure there is not conflated with the `add-permission` result.

File: apps/challenges/views.py [hunk 4]
```diff
@@ -5773,6 +5776,41 @@ def get_challenge_pending_submission_count(request, challenge_pk):
...
+
+    challenges = (
+        Challenge.objects.filter(
+            is_docker_based=True,
+            remote_evaluation=False,
+            approved_by_admin=True,
+        )
+        .exclude(challengeevaluationcluster__isnull=True)
+        .values_list("pk", flat=True)
+    )
```
[the remark this answers is not in the document]
Comment (RishabhJain2018) [reply apps/challenges/views.py:5804]: Right problem, but I took a different fix in 17c80e87.

Excluding disabled challenges from the sweep strands nodes: one disabled while its nodegroup is scaled up would keep those nodes running with nothing left to reconcile it back down — a cost leak, and exactly the drift this sweep exists to catch.

Instead the challenge stays in the sweep and `is_disabled` now forces scale-down, the same path already used for ended challenges. `autoscale_meta` exposes the field and `_should_force_scale_down` returns True on it. Covered by `test_scale_down_when_challenge_disabled`, which asserts `desiredSize=0` for a disabled challenge sitting at 5 nodes with 8 pending submissions.

File: tests/unit/challenges/test_aws_utils.py [hunk 5]
```diff
@@ -3321,6 +3325,15 @@ def test_create_eks_nodegroup_success(
...
         mock_get_boto3_client.assert_called_with("ecs", mock_aws_credentials)
 
+        # The nodegroup name is recorded so autoscaling can target it directly
+        # instead of guessing the first entry from list_nodegroups.
+        mock_evaluation_cluster.objects.filter.assert_called_once_with(
+            challenge_id=mock_challenge.pk
+        )
+        mock_evaluation_cluster.objects.filter.return_value.update.assert_called_once_with(
+            nodegroup_name="Test-Challenge-1-test-env-nodegroup"
+        )
```
[the remark this answers is not in the document]
Comment (RishabhJain2018) [reply tests/unit/challenges/test_aws_utils.py:3576]: Fair, fixed in 17c80e87. Now asserts `create_nodegroup` with its full argument set, so the scaling config, subnets, AMI type, and node role are all pinned rather than just the persistence that follows.

File: scripts/lambda/auto_scale_eks_nodes_lambda.py [hunk 6]
```diff
@@ -225,10 +264,13 @@ def handler(event, context):
...
         scaling_config = {"minSize": 0, "desiredSize": 0, "maxSize": 1}
     else:
+        # maxSize must stay within the challenge's configured cap. Deriving it
+        # from the raw pending count would let a submission burst raise the
+        # ceiling above what the challenge host provisioned for.
         scaling_config = {
             "minSize": 1,
             "desiredSize": target_desired_size,
-            "maxSize": max(target_desired_size, pending_submissions),
+            "maxSize": max(target_desired_size, scale_up_cap),
```
[the remark this answers is not in the document]
Comment (RishabhJain2018) [reply scripts/lambda/auto_scale_eks_nodes_lambda.py:285]: Not taking this one — the suggested expression is behaviourally identical to the current code.

`max(t, min(t, c), c) if t > c else c` reduces to `t if t > c else c`, which is `max(t, c)`. I brute-forced both over `t, c ∈ [0, 12)`: zero differing inputs.

It is also weaker on its own terms. In the `t > c` case it is meant to guard, it returns `t` — the uncapped value the comment warns about. For `t=5, c=2` both expressions yield 5.

The premise does not hold either: `target_desired_size = _desired_size_for_pending(pending, cap) = min(pending, cap)`, so `t > c` is unreachable. The `max()` is not relying on that invariant for correctness — it is there to guarantee `maxSize >= desiredSize`, which the EKS API requires. Leaving as is.

File: scripts/lambda/auto_scale_eks_nodes_lambda.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (RishabhJain2018) [reply scripts/lambda/auto_scale_eks_nodes_lambda.py:390]: Follow-up on this same hunk in 8212a3af: switched `logger.error` to `logger.exception` there. The catch is broad precisely because the exception types landing in it are unanticipated, which is exactly when the traceback is worth keeping in CloudWatch.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by RishabhJain2018** [comment]
Addressed the two outside-diff findings from the latest review in 8212a3af (they had no inline threads to reply on):

**`put_role_policy` / `update_assume_role_policy` asserted only `.called`** — valid, and worth more than a nit here: a role carrying the wrong policy document is indistinguishable from a missing one once autoscaling actually runs. `test_creates_role_and_attaches_policy` now pins `RoleName`, `PolicyName`, `Effect`, and the full sorted EKS action set; `test_refreshes_trust_policy_when_role_already_exists` pins the refreshed trust principal and the policy name. Verified the expected values against `settings.EKS_AUTOSCALE_CROSS_ACCOUNT_POLICY_DOCUMENT` by parsing it directly rather than restating it.

**`update_assume_role_policy.assert_not_called()` in the unexpected-error test** — added. A non-conflict `CreateRole` failure is not the already-exists branch, so the trust policy must not be rewritten.

**Sweep traceback** — `logger.exception` instead of `logger.error`, replied on that thread.

Not taking the duplicated Lambda-auth test boilerplate nitpick (🔵 Trivial). Extracting a mixin means editing `ChallengeAutoscaleInternalAPITest`, which is passing and untouched by this PR, for a purely cosmetic gain. Better as its own cleanup once a third Lambda-authenticated endpoint exists to justify the abstraction.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
